src/http/local_async_normal.hpp

```diff
@@ -254,177 +267,213 @@ struct http_async_connection<Tag, version_major, version_minor>
-                             get_body, callback, placeholders::error,
-                             placeholders::bytes_transferred)),
+              delegate_, request_strand_.wrap([=] (boost::system::error_code const &,
+                                                   std::size_t bytes_transferred) {
+                                                self->handle_received_data(status_message, get_body, callback,
+                                                                           ec, bytes_transferred);
+                                              }),
               bytes_transferred);
-          if (!parsed_ok || indeterminate(parsed_ok)) return;
+          if (!parsed_ok || indeterminate(parsed_ok)) {
+            return;
+          }
         case headers:
           if (ec == boost::asio::error::eof) return;
-          // In the following, remainder is the number of bytes that
-          // remain
-          // in the buffer. We need this in the body processing to make
-          // sure
-          // that the data remaining in the buffer is dealt with before
-          // another call to get more data for the body is scheduled.
-          fusion::tie(parsed_ok, remainder) = this->parse_headers(
+          // In the following, remainder is the number of bytes that remain in
+          // the buffer. We need this in the body processing to make sure that
+          // the data remaining in the buffer is dealt with before another call
+          // to get more data for the body is scheduled.
+          std::tie(parsed_ok, remainder) = this->parse_headers(
               delegate_,
-              request_strand_.wrap(boost::bind(
-                  &this_type::handle_received_data,
-                  this_type::shared_from_this(), headers, get_body, callback,
-                  placeholders::error, placeholders::bytes_transferred)),
+              request_strand_.wrap([=] (boost::system::error_code const &ec,
+                                        std::size_t bytes_transferred) {
+                                     self->handle_received_data(headers, get_body, callback,
+                                                                ec, bytes_transferred);
+                                   }),
               bytes_transferred);
 
-          if (!parsed_ok || indeterminate(parsed_ok)) return;
+          if (!parsed_ok || indeterminate(parsed_ok)) {
+            return;
+          }
 
           if (!get_body) {
-            // We short-circuit here because the user does not
-            // want to get the body (in the case of a HEAD
-            // request).
+            // We short-circuit here because the user does not want to get the
+            // body (in the case of a HEAD request).
             this->body_promise.set_value("");
+            if ( callback )
+              callback( boost::iterator_range<typename std::array<typename char_<Tag>::type, 1024>::const_iterator>(), boost::asio::error::eof );
```

Comment:
Extra spaces at the beginning and end of the invocation. Are these intentional?

Reply:
And here.